### PR TITLE
Process.Close check for null before using variable

### PR DESCRIPTION
**Pull request by matejskubic**
Fix when `Process.Close` throws `NullReferenceException` on `RedirectStandardOutput `or `RedirectStandardError`.

Steps to reproduce:

    process.Start
    process.BeginOutputReadLine
    process.Start
    process.Dispose

Actual result:

    System.NullReferenceException : Object reference not set to an instance of an object.

	Stack Trace:
        System.Diagnostics.Process.Close()
        System.Diagnostics.Process.Dispose(Boolean disposing)

Exception occurs in `Close` method:

```C#
    if (_standardOutput != null && (_outputStreamReadMode == StreamReadMode.AsyncMode || _outputStreamReadMode == StreamReadMode.Undefined))
    {
        if (_outputStreamReadMode == StreamReadMode.AsyncMode)
        {
            _output.CancelOperation(); // <-- _output is null
        }
        _standardOutput.Close();
    }
```

### CODE REVIEW

**Comment by wtgodbe** [comment]
The test failed on Windows10 RS4 with

> System.InvalidOperationException : The Process object must have the UseShellExecute property set to false in order to redirect IO streams.

https://mc.dot.net/#/user/matejskubic/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/28bd645a9ac30b9918204aa66a56de5ebc693a4b/workItem/System.Diagnostics.Process.Tests/analysis/xunit/System.Diagnostics.Tests.ProcessTests~2FStart_RedirectStandardOutput_StartAgainWithoutRedirect

**Comment by wtgodbe** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by wtgodbe** [comment]
The change & test themselves LGTM, but please look into why it failed for NETFX x86 & see if it needs to be disabled for that platform

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs [hunk 1]
```diff
@@ -890,6 +890,9 @@ public void Close()
                     _output = null;
                     _error = null;
 
+                    _pendingOutputRead = false;
+                    _pendingErrorRead = false;
+
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs]: Why reset these?

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs [hunk 2]
```diff
@@ -869,7 +869,7 @@ public void Close()
 
                     if (_standardError != null && (_errorStreamReadMode == StreamReadMode.AsyncMode || _errorStreamReadMode == StreamReadMode.Undefined))
                     {
-                        if (_errorStreamReadMode == StreamReadMode.AsyncMode)
+                        if (_error != null && _errorStreamReadMode == StreamReadMode.AsyncMode)
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs]: These could be done instead just by changing the . to ?. on the invocation below.

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 3]
```diff
@@ -1281,6 +1281,28 @@ public void Start_HasStandardErrorEncodingNonRedirected_ThrowsInvalidOperationEx
             Assert.Throws<InvalidOperationException>(() => process.Start());
         }
 
+        [PlatformSpecific(TestPlatforms.Windows)]
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: Why does the test need to be platform specific?

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 4]
```diff
@@ -1281,6 +1281,28 @@ public void Start_HasStandardErrorEncodingNonRedirected_ThrowsInvalidOperationEx
...
+        [Fact]
+        public void Start_RedirectStandardOutput_StartAgainWithoutRedirect()
+        {
+            var process = new Process
+            {
+                StartInfo = new ProcessStartInfo
+                {
+                    FileName = "cmd",
+                    Arguments = "/C dir",
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: We have helpers for starting processes, used elsewhere in the project, that don't require tying this to specific commands that only work on Windows. It's bee better to use those and make this test more general.

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 5]
```diff
@@ -1281,6 +1281,28 @@ public void Start_HasStandardErrorEncodingNonRedirected_ThrowsInvalidOperationEx
...
+                    RedirectStandardOutput = true,
+                }
+            };
+
+            process.Start();
+            process.BeginOutputReadLine();
+
+            process.Start();
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: Are there things we can assert here, such as the first call returning true and the second returning false? (I've not validated that's actually the case, just speculating.)

**Comment by stephentoub** [comment]
@matejskubic, are you still working on this?  Thanks.

**Comment by karelz** [comment]
@matejskubic ping?

**Comment by stephentoub** [comment]
I pushed commits addressing my feedback.

**Comment by stephentoub** [comment]
> Can you explain why the second call to Start causes things to get into a bad state?

And to answer my own question, the issue is that Process.Start calls Close as the very first thing it does, which in turn resets the state on the instance.

**Pull request merged**